ql/instruments/forwardrateagreement.cpp

```diff
@@ -61,8 +61,72 @@ namespace QuantLib {
+      maturityDate_(index->maturityDate(valueDate)) {
+
+        QL_REQUIRE(notionalAmount > 0.0, "notionalAmount must be positive");
+
+        registerWith(Settings::instance().evaluationDate());
+        registerWith(discountCurve_);
+        registerWith(index_);
+        
+        maturityDate_ = calendar_.adjust(maturityDate_, businessDayConvention_);
+        QL_REQUIRE(valueDate_ < maturityDate_, "valueDate must be earlier than maturityDate");
+        strikeForwardRate_ = InterestRate(strikeForwardRate,
+                                          index->dayCounter(),
+                                          Simple, Once);
+    }
+
+    ForwardRateAgreement::ForwardRateAgreement(const Date& valueDate,
+                                               const Date& maturityDate,
+                                               Position::Type type,
+                                               Rate strikeForwardRate,
+                                               Rate referenceRate,
+                                               Real notionalAmount,
+                                               const DayCounter dayCounter,
+                                               const Calendar fixingCalendar,
+                                               const BusinessDayConvention businessDayConvention,
+                                               Handle<YieldTermStructure> discountCurve)
+    : valueDate_(valueDate), maturityDate_(maturityDate),
+      fraType_(type), notionalAmount_(notionalAmount),
+      dayCounter_(dayCounter), calendar_(fixingCalendar),
+      businessDayConvention_(businessDayConvention),
+      discountCurve_(std::move(discountCurve)),
+      useIndexedCoupon_(false) {
+
+        QL_REQUIRE(notionalAmount > 0.0, "notionalAmount must be positive");
+
+        registerWith(Settings::instance().evaluationDate());
+        registerWith(discountCurve_);
+        
+        maturityDate_ = calendar_.adjust(maturityDate_, businessDayConvention_);
+        QL_REQUIRE(valueDate_ < maturityDate_, "valueDate must be earlier than maturityDate");
+        strikeForwardRate_ = InterestRate(strikeForwardRate,
+                                          dayCounter,
+                                          Simple, Once);
+        referenceRate_ = InterestRate(referenceRate,
+                                      dayCounter,
+                                      Simple, Once);    
+        valueDateTime_ = dayCounter_.yearFraction(Settings::instance().evaluationDate(), valueDate_);
+        maturityDateTime_ = dayCounter_.yearFraction(Settings::instance().evaluationDate(), maturityDate_);                              
+    }
+
     Date ForwardRateAgreement::fixingDate() const {
-        return index_->fixingDate(valueDate_);
+        if (index_)
+            return index_->fixingDate(valueDate_);
+        else // if FRA is constructed without the index it returns the valueDate
```

Comment:
This seems dubious. If i create two FRAs, one without an index, and one without an index but with the exact same value and maturity dates as the one with an index, i would expect them to behave identically. But because of this, they won't - they will have different fixing dates.

Reply:
I agree. I decided to return the `valueDate` here because in case the FRA is created without the index then basically the fixing days can't be implicitly taken from it. Hence I imagined that the `valueDate` should already consider the number of fixing days. If we want to go for the solution where we either take the `IborIndex` or the `Handle<YieldTermStructure>` then I will add one parameter `fixingDays`in the constructor and I will calculate the `fixingDate` accordingly as `valueDate - fixingDays`.